python/lsst/daf/butler/registry/dimensions/table.py

```diff
@@ -0,0 +1,92 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ["TableDimensionRecordStorage"]
+
+from typing import Optional
+
+import sqlalchemy
+
+from ...core import DataCoordinate, DataId, DimensionElement, DimensionRecord
+from ..interfaces import Database, DimensionRecordStorage
+
+
+class TableDimensionRecordStorage(DimensionRecordStorage):
+    """A record storage implementation uses a regular database table.
+
+    For spatial dimension elements, use `SpatialDimensionRecordStorage`
+    instead.
+
+    Parameters
+    ----------
+    db : `Database`
+        Interface to the database engine namespace that will hold these
+        dimension records.
+    element : `DimensionElement`
+        The element whose records this storage will manage.
+    table : `sqlalchemy.schema.Table`
+        The logical table for the element.
+    """
+    def __init__(self, db: Database, element: DimensionElement, *, table: sqlalchemy.schema.Table):
+        self._db = db
+        self._table = table
+        self._element = element
+
+    @property
+    def element(self) -> DimensionElement:
+        # Docstring inherited from DimensionRecordStorage.element.
+        return self._element
+
+    def clearCaches(self):
+        # Docstring inherited from DimensionRecordStorage.clearCaches.
+        pass
+
+    def getElementTable(self, dataId: Optional[DataId] = None) -> Optional[sqlalchemy.sql.FromClause]:
+        return self._table
+
+    def getCommonSkyPixOverlapTable(self, dataId: Optional[DataId] = None
+                                    ) -> Optional[sqlalchemy.sql.FromClause]:
+        return None
+
+    def fetch(self, dataId: DataCoordinate) -> Optional[DimensionRecord]:
+        # Docstring inherited from DimensionRecordStorage.fetch.
+        RecordClass = self.element.RecordClass
+        # I don't know how expensive it is to construct the query below, and
+        # hence how much gain there might be to caching it, so I'm going to
+        # wait for it to appear as a hotspot in a profile before trying that.
+        whereTerms = [self._table.columns[fieldName] == dataId[dimension.name]
+                      for fieldName, dimension in zip(RecordClass.__slots__, self.element.graph.required)]
+        query = sqlalchemy.sql.select(
+            [self._table.columns[name] for name in RecordClass.__slots__]
+        ).select_from(
+            self._table
+        ).where(sqlalchemy.sql.and_(*whereTerms))
+        row = self._db.query(query).fetchone()
+        if row is None:
+            return None
+        return RecordClass(*row)
+
+    def insert(self, *records: DimensionRecord):
```

Comment:
So records is plural but its type is DimensionRecord. Is the type iterable or is the type hint wrong

Reply:
It's a variadic argument, and type annotations for variadic arguments (like those for the type yielded by a generator) are supposed to be for a single element.